src/wpool.erl

```diff
@@ -11,36 +11,180 @@
 % under the License.
-%%% @author Fernando Benavides <elbrujohalcon@inaka.net>
-%%% @doc Worker pool main interface.
-%%%      Use functions provided by this module to manage your pools of workers
+%% @author Fernando Benavides <elbrujohalcon@inaka.net>
+%% @doc Worker pool main interface.
+%%
+%% Use functions provided by this module to manage your pools of workers.
+%%
+%% <h2>Starting the application</h2>
+%% <b>Worker Pool</b> is an Erlang application that can be started using the functions in the
+%% `application' module. For convenience, `wpool:start/0' and `wpool:stop/0' are also provided.
+%%
+%% <h2>Starting a Pool</h2>
+%%
+%% To start a new worker pool, you can either
+%% <ul>
+%%   <li>Use `wpool:child_spec/2' if you want to add the pool under a supervision tree
+%%   initialisation;</li>
+%%   <li>Use `wpool:start_pool/1' or `wpool:start_pool/2' if you want to supervise it yourself;</li>
+%%   <li>Use `wpool:start_sup_pool/1' or `wpool:start_sup_pool/2' if you want the pool to live under
+%%   wpool's supervision tree.</li>
+%% </ul>
+%%
+%% <h2>Stopping a Pool</h2>
+%% To stop a pool, just use `wpool:stop_pool/1' or `wpool:stop_sup_pool/1' according to how you
+%% started the pool.
+%%
+%% <h2>Using the Workers</h2>
+%%
+%% Since the workers are `gen_server's, messages can be `call'ed or `cast'ed to them. To do that you
+%% can use `wpool:call' and `wpool:cast' as you would use the equivalent functions on `gen_server'.
+%%
+%% <h3>Choosing a Strategy</h3>
+%%
+%% Beyond the regular parameters for `gen_server', wpool also provides an extra optional parameter
+%% <b>Strategy</b> The strategy used to pick up the worker to perform the task. If not provided,
+%% the result of `wpool:default_strategy/0' is used.
+%%
+%% The available strategies are defined in the `t:wpool:strategy/0' type.
+%%
+%% <h2>Watching a Pool</h2>
+%% Wpool provides a way to get live statistics about a pool. To do that, you can use
+%% `wpool:stats/1'.
 -module(wpool).
 
 -behaviour(application).
 
-%% Copied from gen.erl
--type debug_flag() :: trace | log | statistics | debug | {logfile, string()}.
--type gen_option() ::
-    {timeout, timeout()} | {debug, [debug_flag()]} | {spawn_opt, [proc_lib:spawn_option()]}.
--type gen_options() :: [gen_option()].
+-type overrun_warning() :: infinity | pos_integer().
+%% The number of milliseconds after which a task is considered <i>overrun</i> i.e., delayed.
+%%
+%% A warning is emitted using {@link overrun_handler()}.
+%%
+%% The task is monitored until it is finished,
+%% thus more than one warning might be emitted for a single task.
+%%
+%% The rounds of warnings are not equally timed, an exponential backoff algorithm is used instead:
+%% after each warning the overrun time is doubled (i.e. with `overrun_warning = 1000' warnings would
+%% be emitted after 1000, 2000, 4000, 8000 ...).
+%%
+%% The default value for this setting is `infinity', i.e., no warnings are emitted.
```

Comment:
So that's how you document types! TIL

Reply:
Now my question is how do I make `rebar3_format` accept it, it fails like this 🥲

Reply:
What happens if you run `rebar3 format`?

Reply:
It puts an empty line between the type declaration and the documentation, and in that case then `ex_doc` thinks is just a comment, instead of documentation 🥲

Reply:
…or a feature request.

Reply:
I bet that Erlang parses this as a pre-comment for the next attribute and that's why rebar3_format adds the inbetween line. Why ex_doc would not read the comment from _before_ the type (as it is the case with functions, modules, etc.) is also… inconsistent, I would say. (¬_¬)

Reply:
Yeah, it is also inconsistent, I'd have expected something like the usual `@doc` tag before the type as it works for before functions... maybe @starbelly has an idea? 🤔

Reply:
😄 This appears to be a bug in edoc when creating chunks, at least at a quick glance. You should able to read in the doc from the shell and then generate the chunk for it. 

```
1> {_Module, Doc, Entries} = edoc:get_doc("src/wpool.erl", [return_entries, private, hidden]).
2> Chunk = edoc:layout(Doc, [{entries, Entries}, {source, "wpool.erl"} | [return_entries, private, hidden]]).
```

You'll notice in the `Doc`, everything is there, but in the `Chunk`, not so, unless there's no break between comment identifiers.  If all of this is write, then we need to send up a fix to OTP. 

Those options are essentially what rebar3_ex_doc uses, `{doclet, edoc_doclet_chunks}`, `{layout, edoc_layout_chunks}` , and `{preprocess, true}`, but this _should_ not have an effect (especially per observing the output via above steps).

Reply:
Small update: Something goes wrong prior to calling `edoc:layout/2`, thus, something I would reckon is going wrong in `edoc_extract`.

Reply:
One more  note 😄  I'm not sure if this is a bug, it seems like a feature 🤔

Reply:
At this point I'm not sure what you're referring to, @starbelly … and I'm too afraid to ask 🙊

Reply:
@NelsonVides pointed me at this thread. Let me try shedding some light on this.

The snippet to generate a chunk has to look like this:

```erlang
{_Module, Doc, Entries} = edoc:get_doc("src/wpool.erl", [return_entries, private, hidden, {doclet, edoc_doclet_chunks}, {layout, edoc_layout_chunks}, {preprocess, true}]).
Chunk = edoc:layout(Doc, [{entries, Entries}, {source, "wpool.erl"}, {doclet, edoc_doclet_chunks}, {layout, edoc_layout_chunks}, {preprocess, true} | [return_entries, private, hidden]]).
% display the structure, not just a binary blob
erlang:binary_to_term(Chunk).
```

@starbelly's snippet generates classic EDoc HTML, despite using the word `Chunk` for the final variable. We have to provide the options specifying which doclet and layout modules to use to generate chunks.

> Why ex_doc would not read the comment from before the type (as it is the case with functions, modules, etc.) is also… inconsistent, I would say. (¬_¬)

ex_doc has nothing to do with the comments themselves. The pipeline looks like this:

```
.erl with EDoc comments -> EDoc (with chunk doclet and layout) -> chunks -> ExDoc -> ExDoc static HTML
```

When ex_doc kicks in, the comments are long gone. ex_doc only cares about the chunk contents.

> I bet that Erlang parses this as a pre-comment for the next attribute and that's why rebar3_format adds the inbetween line.

I don't know why `rebar3_format` adds an extra newline after the `-type` attribute, but if it does, then it's a bug according to the EDoc documentation - see section https://www.erlang.org/doc/apps/edoc/chapter#type-definitions and the examples there:

```erlang
    -type my_list(X) :: [X]. %% A special kind of lists ...

    -opaque another_list(X) :: [X].
    %% another_list() is a kind of list...
```

The canonical form of this is a comment immediately after the attribute, even without a newline. A single newline is also accepted, though. However, any more newlines (like the extra newline possibly added by `rebar3_format`) make this comment detached from the type definition and instead make it a comment attached to the next "meaningful" syntax element (generally, the following function).

Just to make it clear, I personally don't think this format is great and I'm not arguing that nothing better could be invented. On the contrary, I think [`-type` support in EDoc is quite a kludge](https://github.com/erlang/otp/blob/5640a6f58d2fb5e33be588f6bd0f1b0630187ea4/lib/edoc/src/edoc_specs.erl#L137-L159). I just spent some time spelunking EDoc source code when adding chunk generation there, hence I can share some details which are not clear just from the docs. At the very least, though, this format is documented in the EDoc User's Guide (implicitly, by examples, not explicitly by description).

All in all, the expected EDoc format is documented. If some other tool changes the use of this syntax in a source file, then I think a bug fix should be proposed for that tool - in this case `rebar3_format` - unless compatiblity with EDoc is a non-goal.

EDoc has accumulated quite a few kludges over time, so investing time to fix them one by one doesn't make that much sense, given it's original markup is quite outdated and predates many Erlang syntax elements. This is even more true since [new Erlang Markdown docs are just around the corner](https://github.com/erlang/otp/pull/7936).

I hope this helps a little bit 🙏

Reply:
That's a great description of the issue, @erszcz !! Now I understand.
So, yes… it's a formatter bug because when we created the formatter we knew nothing about https://www.erlang.org/doc/apps/edoc/chapter#type-definitions. I'll open a bug report there.
Thank you, very much!

Reply:
AdRoll/rebar3_format#356

Reply:
So… @NelsonVides … either you work on that ticket for the formatter ☝🏻 … or… you _temporarily_ disable the formatting of these modules with `-format ignore.`, put a `@todo remove this line when https://github.com/AdRoll/rebar3_format/issues/356 is fixed.` and move on with the PR.
It's 100% up to you.

Reply:
Right, so for the sake of unblocking this PR I'd prefer a `format ignore` line with the todo. Then when I get some free time again I'll see about getting into the formatter but that I can't promise now, my pipeline for the near future just got filled up 🥲